src/main/scala/com/cognite/spark/datasource/EventsRelationV1.scala

```diff
@@ -0,0 +1,92 @@
+package com.cognite.spark.datasource
+
+import cats.effect.IO
+import com.cognite.sdk.scala.v1.{Event, EventUpdate, GenericClient}
+import com.cognite.sdk.scala.v1.resources.Events
+import com.cognite.spark.datasource.SparkSchemaHelper.{asRow, fromRow, structType}
+import com.softwaremill.sttp.Uri
+import com.softwaremill.sttp._
+import org.apache.spark.sql.{Row, SQLContext}
+import org.apache.spark.sql.sources.InsertableRelation
+import io.circe.generic.auto._
+import org.apache.spark.sql.types.{DataTypes, StructType}
+import cats.implicits._
+
+class EventsRelationV1(config: RelationConfig)(val sqlContext: SQLContext)
+    extends SdkV1Relation[Event, Events[IO], EventItem](config, "events")
+    with InsertableRelation {
+
+  override def insert(rows: Seq[Row]): IO[Unit] = {
+    val events = fromRowWithFilteredMetadata(rows)
+    client.events.createFromRead(events) *> IO.unit
+  }
+
+  override def update(rows: Seq[Row]): IO[Unit] = {
+    val events = rows.map(r => fromRow[EventUpdate](r))
+    client.events.update(events) *> IO.unit
+  }
+
+  override def delete(rows: Seq[Row]): IO[Unit] = {
+    val ids = rows.map(r => fromRow[DeleteItem](r).id)
+    client.events.deleteByIds(ids)
+  }
+
+  override def upsert(rows: Seq[Row]): IO[Unit] = getFromRowAndCreate(rows)
+
+  def fromRowWithFilteredMetadata(rows: Seq[Row]): Seq[Event] =
+    rows.map { r =>
+      val event = fromRow[Event](r)
+      event.copy(metadata = filterMetadata(event.metadata))
+    }
+  override def getFromRowAndCreate(rows: Seq[Row]): IO[Unit] = {
+    val events = fromRowWithFilteredMetadata(rows)
+
+    client.events
+      .createFromRead(events)
+      .handleErrorWith {
+        case e: com.cognite.sdk.scala.common.CdpApiException =>
+          if (e.code == 409) {
+            val existingExternalIds =
+              e.duplicated.get.map(j => j("externalId").get.asString.get)
+            resolveConflict(existingExternalIds, events)
+          } else { IO.raiseError(e) }
+      } *> IO.unit
```

Comment:
I might be wrong here but I think this ignores the possible error from the left hand side, and this should `.map(IO.unit)` instead... @Tapped what do you think?

Reply:
No, it doesn't, an IO product takes the product of the right side if there are no error, however when there are an error it will fail-fast.